eventcore/src/executor.rs

```diff
@@ -674,147 +685,242 @@ where
         for<'a> <C::Event as TryFrom<&'a ES::Event>>::Error: std::fmt::Display,
         ES::Event: From<C::Event> + Clone + serde::Serialize,
     {
-        let mut stream_ids = command.read_streams();
-        let mut stream_resolver = StreamResolver::new();
-        let mut iteration = 0;
-        let max_iterations = options.max_stream_discovery_iterations;
-
+        let context = StreamDiscoveryContext::new(
+            command.read_streams(),
+            options.max_stream_discovery_iterations,
+        );
+        
+        self.execute_with_discovery(command, context, options, StreamResolver::new()).await
+    }
+    
+    /// Recursive functional execution with stream discovery
+    async fn execute_with_discovery<C>(
+        &self,
+        command: C,
+        mut context: StreamDiscoveryContext<stream_discovery::states::Initialized>,
+        options: &ExecutionOptions,
+        mut stream_resolver: StreamResolver,
+    ) -> CommandResult<HashMap<StreamId, EventVersion>>
+    where
+        C: Command,
+        C::Event: Clone + PartialEq + Eq + for<'a> TryFrom<&'a ES::Event> + serde::Serialize,
+        for<'a> <C::Event as TryFrom<&'a ES::Event>>::Error: std::fmt::Display,
+        ES::Event: From<C::Event> + Clone + serde::Serialize,
+    {
+        // Transform recursion into iteration to avoid stack overflow
         loop {
-            iteration += 1;
-            if iteration > max_iterations {
-                return Err(CommandError::ValidationFailed(format!(
-                    "Command '{}' exceeded maximum stream discovery iterations ({max_iterations}). This suggests the command is continuously discovering new streams. Current streams: {:?}",
-                    std::any::type_name::<C>(),
-                    stream_ids.iter().map(std::convert::AsRef::as_ref).collect::<Vec<_>>()
-                )));
-            }
-
+            // Transition to data loaded state
+            let data_loaded_context = context.with_loaded_data();
```

Comment:
I thought we were going to get rid of the looping construct in favor of a functional iteration approach. Can you explain why we're unable to do that? Or did this just get missed?